### PR TITLE
[SPARK-45575][SQL] Support time travel options for df read API

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

We've added time travel API in DS v2 and a dedicated SQL syntax for it. However, there is no way to do it with df read API. This PR adds time travel options (`timestampAsOf` and `versionAsOf`) to support time travel with df read API.

### Why are the changes needed?

feature parity 

### Does this PR introduce _any_ user-facing change?

Yes, now people can specify time travel in read options.

### How was this patch tested?

new tests

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

File: common/utils/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -2196,6 +2208,13 @@
     },
     "sqlState" : "42K0E"
   },
+  "INVALID_TIME_TRAVEL_TIMESTAMP_OPTION" : {
```
Comment (yaooqinn) [inline common/utils/src/main/resources/error/error-classes.json:2211]: Duplicates INVALID_TIME_TRAVEL_TIMESTAMP_EXPR.INPUT?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TimeTravelSpec.scala [hunk 2]
```diff
@@ -65,4 +66,36 @@ object TimeTravelSpec {
...
+    if (timestampStr != null && versionStr != null) {
+      throw QueryCompilationErrors.invalidTimeTravelSpecError()
+    }
+
+    if (timestampStr != null) {
+      val timestampValue = Cast(
+        Literal(timestampStr),
+        TimestampType,
+        Some(sessionLocalTimeZone),
+        ansiEnabled = false
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TimeTravelSpec.scala:86]: We intentionally turn off ansi here so it doesn't fail, and we can customize the error message later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TimeTravelSpec.scala [hunk 3]
```diff
@@ -65,4 +66,35 @@ object TimeTravelSpec {
...
+        val timestampValue = Cast(
+          Literal(timestampStr),
+          TimestampType,
+          Some(sessionLocalTimeZone),
+          ansiEnabled = false
+        ).eval()
+        if (timestampValue == null) {
+          throw new AnalysisException(
+            "INVALID_TIME_TRAVEL_TIMESTAMP_EXPR.OPTION",
+            Map("expr" -> s"'$timestampStr''")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TimeTravelSpec.scala:89]:
```suggestion
            Map("expr" -> s"'$timestampStr'")
```

File: common/utils/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -2188,6 +2194,11 @@
           "Must be deterministic."
         ]
       },
+      "OPTION" : {
+        "message" : [
+          "Timestamp string in the options should be in the format of 'yyyy-MM-dd HH:mm:ss[.us][zone_id]'."
```
Comment (yaooqinn) [inline common/utils/src/main/resources/error/error-classes.json:2199]: The given formatter might be insufficient for valid timestamps and the `[.us][zone_id]` part differs from https://spark.apache.org/docs/latest/sql-ref-datetime-pattern.html.

**Review by ryan-johnson-databricks — APPROVED** [review]
LGTM, tho we should resolve the timestamp format question the other reviewer raised
(I'm not an expert there, but it sounds important)

File: common/utils/src/main/resources/error/error-classes.json (hunk 4 above)
Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-classes.json:2199]: We use Cast and the allowed input string is very flexible. But in the document, we should ask people to use the most standard timestamp format.

BTW, this is not really a Java timestamp format syntax, but only for document purposes. I can remove the `[.us][zone_id]` part as most people won't specify them in time travel.

File: common/utils/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -2207,6 +2213,11 @@
           "Must be deterministic."
         ]
       },
+      "OPTION" : {
+        "message" : [
+          "Timestamp string in the options should be in the format of 'yyyy-MM-dd HH:mm:ss'."
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-classes.json:2218]:
```suggestion
          "Timestamp string in the options should use ISO standard format such as 'yyyy-MM-dd HH:mm:ss'."
```

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 6]
```diff
@@ -2928,29 +2932,34 @@ class DataSourceV2SQLSuiteV1Filter
...
+      assert(res6 === Array(Row(7), Row(8)))
+      val res7 = sql("SELECT * FROM t TIMESTAMP AS OF make_date(2021, 1, 29)").collect()
+      assert(res7 === Array(Row(7), Row(8)))
+      val res8 = sql("SELECT * FROM t TIMESTAMP AS OF to_timestamp('2021-01-29 00:00:00')").collect()
+      assert(res8 === Array(Row(7), Row(8)))
       // Scalar subquery is also supported.
-      assert(sql("SELECT * FROM t TIMESTAMP AS OF (SELECT make_date(2021, 1, 29))").collect()
-        === Array(Row(7), Row(8)))
+      val res9 = sql("SELECT * FROM t TIMESTAMP AS OF (SELECT make_date(2021, 1, 29))")
+      assert(res9.collect() === Array(Row(7), Row(8)))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:2958]:
```suggestion
      val res9 = sql("SELECT * FROM t TIMESTAMP AS OF (SELECT make_date(2021, 1, 29))").collect()
      assert(res9 === Array(Row(7), Row(8)))
```

File: docs/sql-error-conditions-invalid-time-travel-timestamp-expr-error-class.md [hunk 7]
```diff
@@ -33,6 +33,10 @@ Cannot be casted to the "TIMESTAMP" type.
 
 Must be deterministic.
 
+## OPTION
+
+Timestamp string in the options should be in the format of 'yyyy-MM-dd HH:mm:ss[.us][zone_id]'.
```
Comment (yaooqinn) [inline docs/sql-error-conditions-invalid-time-travel-timestamp-expr-error-class.md:38]:
```suggestion
Timestamp string in the options should use ISO standard format such as 'yyyy-MM-dd HH:mm:ss'.
```

**Review by yaooqinn — APPROVED** [review]

File: common/utils/src/main/resources/error/error-classes.json [hunk 8]
```diff
@@ -2207,6 +2213,11 @@
           "Must be deterministic."
         ]
       },
+      "OPTION" : {
+        "message" : [
+          "Timestamp string in the options should use ISO standard format such as 'yyyy-MM-dd HH:mm:ss'."
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:2218]: > 

Which ISO standard do you mean? As per [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) :
In ISO 8601:2004 it was permitted to omit the "T" character by mutual agreement as in "200704051430",[[36]](https://en.wikipedia.org/wiki/ISO_8601#cite_note-36) **but this provision was removed in ISO 8601-1:2019**. Separating date and time parts with other characters such as space is not allowed in ISO 8601, but allowed in its profile RFC 3339.

Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-classes.json:2218]: hmm, shall we say Java standard instead? `yyyy-MM-dd HH:mm:ss` is pretty common.

Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:2218]: Yep

Comment (yaooqinn) [reply common/utils/src/main/resources/error/error-classes.json:2218]: Is it actually ANSI SQL Standard DATE/TIMESTAMP and TIMESTAMP WITH TIME ZONE formats?

**Review by MaxGekk — APPROVED** [review]

File: common/utils/src/main/resources/error/error-classes.json (hunk 8 above)
Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-classes.json]: does ANSI define timestamp format using these characters `yyyy-MM-dd HH:mm:ss`?

Comment (yaooqinn) [reply common/utils/src/main/resources/error/error-classes.json]: Yes，it's defined in the ANSI SQL Standard in BNF form
```bnf
<unquoted timestamp string> ::=
  <unquoted date string> <space> <unquoted time string>

```
image

Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-classes.json]: hmm, it's probably too much to put this BNF in the error message. `yyyy-MM-dd HH:mm:ss` from Java looks clearer.

Comment (yaooqinn) [reply common/utils/src/main/resources/error/error-classes.json]: I remember you once told me try our best to avoid Java stuff from SQL docs and errors:)

Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-classes.json]: alright, how about `... use ISO standard format 'yyyy-MM-ddTHH:mm:ss'`

Comment (cloud-fan) [reply common/utils/src/main/resources/error/error-classes.json]: I believe `T` is the valid delimiter to separate date and time in ISO standard.

Comment (yaooqinn) [reply common/utils/src/main/resources/error/error-classes.json]: How about
```
Timestamp string in the options should be able to cast to TIMESTAMP.
```

Refering INVALID_TIME_TRAVEL_TIMESTAMP_EXPR.INPUT, it seems that we also do a cast operation here

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
Thank you all, merged to master

**Pull request closed without merging**